### PR TITLE
Introduce a repair scan to fix failing clusters

### PR DESCRIPTION
**Pull request by alexeyklyukin**
A repair is a sync scan that acts only on those clusters that indicate
that the last add, update or sync operation on them has failed. It is
supposed to kick in more frequently than the repair scan. The repair
scan still remains to be useful to fix the consequences of external
actions (i.e. someone deletes a postgres-related service by mistake)
unbeknownst to the operator.

The repair scan is controlled by the new repair_period parameter in the
operator configuration. It has to be at least 2 times more frequent than
a sync scan to have any effect (a normal sync scan will update both last
synced and last repaired attributes of the controller, since repair is
just a sync underneath).

A repair scan could be queued for a cluster that is already being synced
if the sync period exceeds the interval between repairs. In that case a
repair event will be discarded once the corresponding worker finds out
that the cluster is not failing anymore.

### CODE REVIEW

**Comment by zerg-junior** [comment]
:+1:

**Comment by alexeyklyukin** [comment]
👍

File: docs/index.md [hunk 1]
```diff
@@ -19,6 +19,10 @@ manages PostgreSQL clusters on Kubernetes:
 
 3. Finally, the operator periodically synchronizes the actual state of each
    Postgres cluster with the desired state defined in the cluster's manifest.
+   This is triggered by either the `sync scan`, running every `resync_period`
+   seconds for every cluster, or by the `repair scan`, coming every
+   `repair_period` only for those clusters that didn't report success as a
+   result of the last operation running on them.
```
Comment (zerg-junior) [inline docs/index.md]: This paragraph  is better fitted for the administrator docs; I wrote the Intro with the intention to provide a very high-level overview of operator's capabilities w/o any overly technical details.

File: docs/reference/operator_parameters.md [hunk 2]
```diff
@@ -80,7 +80,10 @@ Those are top-level keys, containing both leaf keys and groups.
   are applied. The default is `-1`.
 
 * **resync_period**
-  period between consecutive sync requests. The default is `5m`.
+  period between consecutive sync requests. The default is `30m`.
+
+* **repair_period**
+  period between consecutive repair requests. The default is `5m`.
```
Comment (zerg-junior) [inline docs/reference/operator_parameters.md]: it is probably worth to mention here a shortened version [of this pr descriotion](https://github.com/zalando-incubator/postgres-operator/pull/304#issue-190550137)

File: pkg/controller/postgresql.go [hunk 3]
```diff
@@ -41,9 +41,14 @@ func (c *Controller) clusterResync(stopCh <-chan struct{}, wg *sync.WaitGroup) {
 
 // TODO: make a separate function to be called from InitSharedInformers
 // clusterListFunc obtains a list of all PostgreSQL clusters and runs sync when necessary
+// NB: as this function is called directly by the informer, it needs to avoid acquiring locks
+// on individual cluster structures. Therefore, it acts on the maifests obtained from Kubernetes
```
Comment (zerg-junior) [inline pkg/controller/postgresql.go]: maNifests :)

File: pkg/controller/postgresql.go [hunk 4]
```diff
@@ -60,19 +65,40 @@ func (c *Controller) clusterListFunc(options metav1.ListOptions) (runtime.Object
...
+	timeFromPreviousRepair := currentTime - atomic.LoadInt64(&c.lastClusterRepairTime)
+	if timeFromPreviousSync >= int64(c.opConfig.ResyncPeriod.Seconds()) {
+		event = spec.EventSync
+	} else if timeFromPreviousRepair >= int64(c.opConfig.RepairPeriod.Seconds()) {
+		event = spec.EventRepair
 	}
+	if event != "" {
+		c.queueEvents(&list, event)
+	} else {
+		c.logger.Infof("not enough passed since the last sync (%s seconds) or repair (%s seconds)", timeFromPreviousSync, timeFromPreviousRepair)
```
Comment (zerg-junior) [inline pkg/controller/postgresql.go]: not enough *time* ?

File: pkg/controller/postgresql.go [hunk 5]
```diff
@@ -170,6 +201,16 @@ func (c *Controller) processEvent(event spec.ClusterEvent) {
...
 	defer c.curWorkerCluster.Store(event.WorkerID, nil)
 
+	if event.EventType == spec.EventRepair {
+		runRepair, lastOperationStatus := cl.NeedsRepair()
+		if !runRepair {
+			lg.Debugf("Observed cluster status %s, repair is not required", lastOperationStatus)
+			return
+		}
+		lg.Debugf("Observed cluster status %s, running sync scan to repair the cluster", lastOperationStatus)
+		event.EventType = spec.EventSync
```
Comment (zerg-junior) [inline pkg/controller/postgresql.go]: so this is the "under-the-hood" point where the repairs scan turn into a sync scan ?

File: docs/index.md (hunk 1 above)
Comment (alexeyklyukin) [reply docs/index.md]: Good point, moved to the end of the admin guide

File: docs/reference/operator_parameters.md (hunk 2 above)
Comment (alexeyklyukin) [reply docs/reference/operator_parameters.md]: The parameter reference doesn't serve the goal of explaining underline concepts behind those parameters.

File: pkg/controller/postgresql.go (hunk 5 above)
Comment (alexeyklyukin) [reply pkg/controller/postgresql.go]: Right, once the operator verifies the actual cluster status in the operator memory indicates the need for the repair actions it continues with the sync scan.

**Comment by alexeyklyukin** [comment]
👍

File: pkg/controller/postgresql.go [hunk 6]
```diff
@@ -60,19 +65,41 @@ func (c *Controller) clusterListFunc(options metav1.ListOptions) (runtime.Object
...
+	if event != "" {
+		c.queueEvents(&list, event)
+	} else {
+		c.logger.Infof("not enough time passed since the last sync (%s seconds) or repair (%s seconds)",
+			timeFromPreviousSync, timeFromPreviousRepair)
+	}
+	return &list, err
+}
 
+// queueSyncEvents adds a sync event for every cluster with the valid manifest to the queue.
```
Comment (zerg-junior) [inline pkg/controller/postgresql.go]: leftover comment ? afaik there is no `queueSyncEvent` in this file

**Comment by zerg-junior** [comment]
:+1:

File: pkg/controller/postgresql.go (hunk 6 above)
Comment (alexeyklyukin) [reply pkg/controller/postgresql.go]: thanks, fixed

**Comment by alexeyklyukin** [comment]
👍

**Comment by zerg-junior** [comment]
:+1:

**Pull request merged**